chapter_preliminaries/linear-algebra.ipynb

```diff
@@ -0,0 +1,1881 @@
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "NDArray B = nd.ones(new Shape(4,3));\n",
+    "A.dot(B)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 108
+   },
+   "source": [
+    "Matrix-matrix multiplication can be simply called *matrix multiplication*, and should not be confused with the Hadamard product.\n",
+    "\n",
+    "\n",
+    "## Norms\n",
+    ":label:`subsec_lin-algebra-norms`\n",
+    "\n",
+    "Some of the most useful operators in linear algebra are *norms*.\n",
+    "Informally, the norm of a vector tells us how *big* a vector is.\n",
+    "The notion of *size* under consideration here\n",
+    "concerns not dimensionality\n",
+    "but rather the magnitude of the components.\n",
+    "\n",
+    "In linear algebra, a vector norm is a function $f$ that maps a vector\n",
+    "to a scalar, satisfying a handful of properties.\n",
+    "Given any vector $\\mathbf{x}$,\n",
+    "the first property says\n",
+    "that if we scale all the elements of a vector\n",
+    "by a constant factor $\\alpha$,\n",
+    "its norm also scales by the *absolute value*\n",
+    "of the same constant factor:\n",
+    "\n",
+    "$$f(\\alpha \\mathbf{x}) = |\\alpha| f(\\mathbf{x}).$$\n",
+    "\n",
+    "\n",
+    "The second property is the familiar triangle inequality:\n",
+    "\n",
+    "$$f(\\mathbf{x} + \\mathbf{y}) \\leq f(\\mathbf{x}) + f(\\mathbf{y}).$$\n",
+    "\n",
+    "\n",
+    "The third property simply says that the norm must be non-negative:\n",
+    "\n",
+    "$$f(\\mathbf{x}) \\geq 0.$$\n",
+    "\n",
+    "That makes sense, as in most contexts the smallest *size* for anything is 0.\n",
+    "The final property requires that the smallest norm is achieved and only achieved\n",
+    "by a vector consisting of all zeros.\n",
+    "\n",
+    "$$\\forall i, [\\mathbf{x}]_i = 0 \\Leftrightarrow f(\\mathbf{x})=0.$$\n",
+    "\n",
+    "You might notice that norms sound a lot like measures of distance.\n",
+    "And if you remember Euclidean distances\n",
+    "(think Pythagoras' theorem) from grade school,\n",
+    "then the concepts of non-negativity and the triangle inequality might ring a bell.\n",
+    "In fact, the Euclidean distance is a norm:\n",
+    "specifically it is the $L_2$ norm.\n",
+    "Suppose that the elements in the $n$-dimensional vector\n",
+    "$\\mathbf{x}$ are $x_1, \\ldots, x_n$.\n",
+    "The $L_2$ *norm* of $\\mathbf{x}$ is the square root of the sum of the squares of the vector elements:\n",
+    "\n",
+    "$$\\|\\mathbf{x}\\|_2 = \\sqrt{\\sum_{i=1}^n x_i^2},$$\n",
+    "\n",
+    "where the subscript $2$ is often omitted in $L_2$ norms, i.e., $\\|\\mathbf{x}\\|$ is equivalent to $\\|\\mathbf{x}\\|_2$. In code, we can calculate the $L_2$ norm of a vector as follows.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 46,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public NDArray l2Norm(NDArray w){\n",
```

Comment:
I created this simple function because I didn't find anything similar already created in DJL. If there is indeed a function that does the same thing, please let me know so I can use the built-in function.

Reply:
we don't but we can add it to NDArray class like this one https://github.com/awslabs/djl/commit/335556fe7a464df1485201f3fe5e2fb2bdc6b5c0